input/posts/2020-07-08-grpc-reflection-in-net.md

```diff
@@ -0,0 +1,226 @@
+
+Now that we know the name of the services, we can list the methods.
+```cmd
+dotnet grpc-cli ls https://localhost:5001 Reflection.Sample.Greeter
+
+filename: Reflection.Sample.Greeter.proto
+package: Reflection.Sample
+service Greeter {
+  rpc SayHello(Reflection.Sample.HelloRequest) returns (Reflection.Sample.HelloReply) {}
+}
+```
+
+And also dump the output in .proto format
+```cmd
+dotnet grpc-cli dump https://localhost:5001 Reflection.Sample.Greeter
+
+---
+File: Reflection.Sample.HelloRequest.proto
+---
+syntax = "proto3";
+package Reflection.Sample;
+
+message HelloRequest {
+  string Name = 1;
+}
+
+---
+File: Reflection.Sample.HelloReply.proto
+---
+syntax = "proto3";
+package Reflection.Sample;
+
+message HelloReply {
+  string Message = 1;
+}
+
+---
+File: Reflection.Sample.Greeter.proto
+---
+syntax = "proto3";
+import "Reflection.Sample.HelloRequest.proto";
+import "Reflection.Sample.HelloReply.proto";
+package Reflection.Sample;
+
+service Greeter {
+   rpc SayHello(HelloRequest) returns (HelloReply);
+}
+```
+
+We probably want this written to the filesystem, so we will then run the following.
+```cmd
+dotnet grpc-cli dump https://localhost:5001 Reflection.Sample.Greeter -o ./protos
+```
+
+That's it! We now have .proto files, generated from our server reflection endpoint. These .proto files can now be used to generate client stubs in the language of your choice.
+
+### Summary
+As you can see gRPC server reflection can be very useful in code-first scenarios, but it's also useful for testing and debugging. E.g. gRPC CLI supports invoking methods, and this is something I'm also planning on supporting in the .NET global tool.
+
+Code-first using protobuf-net.Grpc is in my opinion the simplest, and most straightforward way to port over your old WCF services to .NET Core and gRPC. It's way more simpler than the official guide over at [Microsoft Docs](https://docs.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/migrate-wcf-to-grpc) which uses Google's protobuf instead of protobuf-net. With protobuf-net.Grpc, you'll most likely get to keep all your data contracts intact, while only needing to add the `Order` property to your data member attributes. Your service contracts, however, will need some rework. But in the end, these are minimal changes compared to porting your data contracts and service contracts over to .proto format. If you also need interoperability from other programming languages, it's a peace of cake to generate .proto files from your gRPC service. So, what are you waiting for? Start porting over your WCF services to .NET Core now, and make sure to use protobuf-net.Grpc when you do it.
```

Comment:
(no specific location) one other thing that may be worth mentioning: [`protogen`](https://www.nuget.org/packages/protobuf-net.Protogen/) also now has the ability to generate C# from these endpoints via the new `--grpc` option, but: it is new and slightly experimental